File: money-core/src/main/scala/com/comcast/money/core/Span.scala
package com.comcast.money.core

import scala.collection.Map

case class Span(
  spanId: SpanId,
  spanName: String,
  appName: String,
  host: String,
  startTime: Long,
  success: Boolean,
  duration: Long,
  notes: Map[String, Note[_]]
  )

File: money-basic/src/main/java/com/comcast/money/basic/Span.java
package com.comcast.money.basic;

public interface Span {

    void start(Long startTime, Span parentSpan, boolean propagate);

    /**
     * Ends a span, moving it to a Closing state
     * @param stopTime The time when the span ended (in milliseconds)
     * @param result The result of the span (success or failure)
     */
    void stop(Long stopTime, boolean result);

    /**
     * Different than end, this will shut down the span and have it emit it's data
     */
    void close();

    void record(Note<?> note);

    void startTimer(String timerKey, Long startTime);

    void stopTimer(String timerKey, Long endTime);

    void timedOut();

    SpanData data();

    boolean isTimedOut();

    boolean shouldClose();
}
